File: ruby/21.merge-two-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}
def merge_two_lists(list1, list2)
  return nil if list1.nil? && list2.nil?

  dummy = ListNode.new

  # 最後に先頭のnodeを返すためのもの
  # 同じオブジェクトを共有している
  beginning = dummy

  while list1 && list2
    if list1.val > list2.val
      dummy.next = list2
      list2 = list2.next
    else
      dummy.next = list1
      list1 = list1.next
    end
    dummy = dummy.next
  end

  if list1
    dummy.next = list1
  elsif list2
    dummy.next = list2
  end
  beginning.next
end

File: rust/src/21.merge-two-sorted-lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>,
        mut list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut beginning = &mut dummy;

        // while let (Some(n1), Some(n2)) = (Some(l1), Some(l2)) {
        while list1.is_some() && list2.is_some() {
            if list1.as_ref().unwrap().val < list2.as_ref().unwrap().val {
                let next = list1.as_mut().unwrap().next.take();
                beginning.next = list1.take();
                list1 = next;
            } else {
                let next = list2.as_mut().unwrap().next.take();
                beginning.next = list2.take();
                list2 = next;
            }
            beginning = beginning.next.as_mut().unwrap();
        }
        if list1.is_some() {
            beginning.next = list1;
        } else {
            beginning.next = list2;
        }
        dummy.next
    }
}
